### DISCUSSION TITLE
Single pyproject.toml package configuration file

### DISCUSSION THREAD

**Discussion by seb5g** [PyMoDAQ/PyMoDAQ#423] [Ideas]
Starting with pymodaq 5, building pymodaq and plugins will use [hatch](https://hatch.pypa.io/latest/) and the [pyproject.toml](https://packaging.python.org/en/latest/guides/writing-pyproject-toml/) configuration file.

For this, I produced a pymodaq_plugins_template branch called 5.0.x that will include this mechanism together with usage of pymodaq 5 and its new submodules: pymodaq_utils, _data and _gui.

To summarize what should be done in a new flavor plugin or to move from v4 to v5

* make sure the pyproject.toml contains all entries as specified in the template
* make sure to use imports from pymodaq_utils, _data. It is not a big requirement as most of the functionalities can still be imported from pymodaq itself (itself importing these from their dedicated module)
* make sure the project structure contains the file called: hatch_build.py

**Comment by seb5g** [comment]
The documentation therefore need to be modified... wherever appears the plugin_info stuff... who could do the job to update the doc in v5?

**Reply by seb5g** [reply]
The pymodaq_plugins_mock last version is using this new template

**Comment by quantumm** [comment]
If I understand correctly, you are talking about writing the documentation about _How to modify your plugin from v4 to v5?_ for the maintainers?

**Reply by seb5g** [reply]
well there is this, that should appear in both doc v4 and v5

and then in the v5 branch: all the mention of the plugin_info, setup.py should be changed to talk about the usage of a single pyproject.toml (I'm not sure of where are all the mention of this...)

**Reply by quantumm** [reply]
Hmmm it sounds like a shitty task about the documentation, that is for me ;)

**Reply by seb5g** [reply]
lol

**Reply by loicguilmard** [reply]
don't worry @quantumm i started some of the modification of the documentation to take into account the changes of the v5. I can look into the  pyporject.toml if you want.
I am not sure i really know what are all the changes though.
Is it just changing the syntax of the parameter and remove the SHORT_PLUGIN_NAME ?
Are we writing down a hatch tutorial too ?
I'll fork the mock project and play around to check the requierements.

**Reply by seb5g** [reply]
both the mock and the template should have the single pyproject file ready. You can look into it!

What about hatch? environment? builds?

**Reply by loicguilmard** [reply]
i am looking into the hatch env part. 
And i am started writing the documentation for the migration plugings from 4.x to 5.x.
I don't think we can generate msi directly from hatch but the env management from hatch would eliminate most of the "tech gap".

**Reply by seb5g** [reply]
I looked a bit into the msi part. Hatch itself is now shipped with a msi, maybe we could ask the maintainer how he did it? In ours, we would add the hatch installation, environment creation and pymodaq installation, then run in it the plugin manager for people to install their plugins

**Comment by quantumm** [comment]
Ok I leave it to you then thanks ;)
I guess a tutorial about hatch would be nice too!
